api.go

```diff
@@ -697,6 +705,39 @@ func (r *Raft) tryRestoreSingleSnapshot(snapshot *SnapshotMeta) bool {
 	return true
 }
 
+// recoverFromCommittedLogs recovers the Raft node from committed logs.
+func (r *Raft) recoverFromCommittedLogs() error {
+	if !r.fastRecovery {
+		return nil
+	}
+
+	// If the store implements CommitTrackingLogStore, we can read the commit index from the store.
+	// This is useful when the store is able to track the commit index and we can avoid replaying logs.
+	store, ok := r.logs.(CommitTrackingLogStore)
+	if !ok {
+		return nil
+	}
+
+	commitIndex, err := store.GetCommitIndex()
+	if err != nil {
+		r.logger.Error("failed to get commit index from store", "error", err)
+		panic(err)
+	}
+
+	lastIndex, err := r.logs.LastIndex()
+	if err != nil {
+		r.logger.Error("failed to get last log index from store", "error", err)
+		panic(err)
+	}
```

Comment:
Shouldn't we fallback to the non fast recovery path instead of panic?

Reply:
🤔 Great question.

For transient errors it's probably better to hard stop rather than silently weaken the expected guarantees. Someone who has enabled this feature (and provided a supported logstore) should be able to assume that by the time raft has started (without error), their FSM is at least as up-to-date as it was before a restart. Silently falling back seems like it makes it impossible to really trust that guarantee and may mean similar bugs that we are trying to prevent are still possible with no way to mitigate them (even if they are rarer).

In most use cases where each server is using this library as a core of it's functionality (i.e. all our products) and only uses a single raft group initialised during startup, a panic is probably reasonable too because then we crash and the supervisor will restart and if it was a transient error then great and if not it's no different to what will happen later when we are unable to read logs.

That said, in writing this I realised it's probably too strong of a decision to make in a library here: it would be possible for a server process to manage multiple raft instances for example and a fatal error in one of them shouldn't terminate the whole process. So I think I'd vote for making these cases return a hard error from `NewRaft` like we do if we fail to read a log or snapshot during startup rather than panic.

Reply:
I guess returning an error from `NewRaft` would allow the library user to decide if they would like to do a fallback or fail starting the application, that is reasonable I think.

I think we should implement this as a specific error type that we publish on the API to allow calling `errors.Is` on it without relying on string compare.